File: stat.py
# python stat.py '{"Func": "Levene", "Samples": [[7, 14, 14, 13, 12, 9, 6, 14, 12, 8], [15, 17, 13, 15, 15, 13, 9, 12, 10, 8], [6, 8, 8, 9, 5, 14, 13, 8, 10, 9]]}'
import collections
import logging
import json
import sys

import numpy as np
import scipy.stats as stats


def pingouin_welch(samples):
    stacked = []
    for i, sample in enumerate(samples):
        for s in sample:
            stacked.append([i, s])
    stacked = np.array(stacked)
    df = pd.DataFrame(stacked, columns=["sample", "v"])

    aov = pingouin.welch_anova(dv="v", between="sample", data=df)

    return aov.iloc[0]


def welch(samples):
    stacked = []
    for i, sample in enumerate(samples):
        for s in sample:
            stacked.append([i, s])
    stacked = np.array(stacked)
    df = pd.DataFrame(stacked, columns=["sample", "v"])

    data = df
    dv = "v"
    between = "sample"

    # Number of groups
    r = data[between].nunique()
    ddof1 = r - 1

    # Compute weights and ajusted means
    grp = data.groupby(between, observed=True, group_keys=False)[dv]
    weights = grp.count() / grp.var()
    adj_grandmean = (weights * grp.mean()).sum() / weights.sum()
    logging.info("weights: %s, adj_grandmean: %s", weights, adj_grandmean)

    # Sums of squares (regular and adjusted)
    ss_res = grp.apply(lambda x: (x - x.mean()) ** 2).sum()
    ss_bet = ((grp.mean(numeric_only=True) - data[dv].mean()) ** 2 * grp.count()).sum()
    logging.info("grp.mean: %s, data.mean: %s, grp.count: %s", grp.mean(numeric_only=True), data[dv].mean(), grp.count())
    ss_betadj = np.sum(weights * np.square(grp.mean(numeric_only=True) - adj_grandmean))
    ms_betadj = ss_betadj / ddof1
    logging.info("ss_res: %s, ss_bet: %s, ss_betadj: %s, ms_betadj: %s", ss_res, ss_bet, ss_betadj, ms_betadj)

    # Calculate lambda, F-value, p-value and np2
    lamb = (3 * np.sum((1 / (grp.count() - 1)) * (1 - (weights / weights.sum())) ** 2)) / (
        r**2 - 1
    )
    logging.info("lamb0: %s, cnt: %s, weighs: %s, weightsSum: %s", np.sum((1 / (grp.count() - 1)) * (1 - (weights / weights.sum())) ** 2), grp.count(), weights, weights.sum())
    fval = ms_betadj / (1 + (2 * lamb * (r - 2)) / 3)
    logging.info("lamb: %s, fval: %s", lamb, fval)
    pval = stats.f.sf(fval, ddof1, 1 / lamb)
    np2 = ss_bet / (ss_bet + ss_res)

    # Create output dataframe
    aov = pd.DataFrame(
        {
            "Source": between,
            "ddof1": ddof1,
            "ddof2": 1 / lamb,
            "F": fval,
            "p-unc": pval,
            "np2": np2,
        },
        index=[0],
    )

    return aov.iloc[0]


LeveneResult = collections.namedtuple('LeveneResult', ('statistic', 'pvalue'))


def levene(*samples, center='median', proportiontocut=0.05):
    k = len(samples)
    Ni = np.empty(k)
    Yci = np.empty(k, 'd')
    def func(x):
        return np.median(x, axis=0)

    for j in range(k):
        Ni[j] = len(samples[j])
        Yci[j] = func(samples[j])
    Ntot = np.sum(Ni, axis=0)
    logging.info("k: %s, Ni: %s, Yci: %s, Ntot %s", k, Ni, Yci, Ntot)

    # compute Zij's
    Zij = [None] * k
    for i in range(k):
        Zij[i] = abs(np.asarray(samples[i]) - Yci[i])

    # compute Zbari
    Zbari = np.empty(k, 'd')
    Zbar = 0.0
    for i in range(k):
        Zbari[i] = np.mean(Zij[i], axis=0)
        Zbar += Zbari[i] * Ni[i]

    Zbar /= Ntot
    numer = (Ntot - k) * np.sum(Ni * (Zbari - Zbar)**2, axis=0)
    logging.info("ZBari: %s, Zbar: %s, numer: %s", Zbari, Zbar, numer)

    # compute denom_variance
    dvar = 0.0
    for i in range(k):
        dvar += np.sum((Zij[i] - Zbari[i])**2, axis=0)

    denom = (k - 1.0) * dvar
    logging.info("dvar: %s, denom: %s", dvar, denom)

    W = numer / denom
    pval = stats.distributions.f.sf(W, k-1, Ntot-k)  # 1 - cdf
    return LeveneResult(W, pval)


def main():
    logging.basicConfig()
    lg = logging.getLogger()
    lg.setLevel(logging.INFO)
    lg.handlers[0].setFormatter(logging.Formatter("%(asctime)s.%(msecs)03d %(pathname)s:%(lineno)d %(message)s", datefmt="%Y-%m-%d %H:%M:%S"))

    cfg = json.loads(sys.argv[1])
    if cfg["Func"] == "Levene":
        scipyOut = stats.levene(*cfg["Samples"])
        # scipyOut = levene(*cfg["Samples"])
        out = {"Levene": {"Statistic": scipyOut.statistic, "PValue": scipyOut.pvalue}}
    elif cfg["Func"] == "Welch":
        global pd, pingouin
        import pandas as pd
        import pingouin
        pingouinOut = pingouin_welch(cfg["Samples"])
        # pingouinOut = welch(cfg["Samples"])
        out = {"Welch": {"F": pingouinOut["F"], "PValue": pingouinOut["p-unc"]}}
    elif cfg["Func"] == "Holm":
        global pd, pingouin
        import pandas as pd
        import pingouin
        reject, pvalsCorr = pingouin.multicomp(cfg["Samples"][0], method="holm")
        out = {"Holm": {"Reject": reject.tolist(), "PValuesCorrected": pvalsCorr.tolist()}}
    else:
        logging.info("unknown function \"%s\"", cfg["Func"])
    print(json.dumps(out))


if __name__ == "__main__":
    main()

File: stat.go
// Package stat provides functions for statistical analysis.
package stat

import (
	"math"

	"github.com/aclements/go-moremath/stats"
	"gonum.org/v1/gonum/stat/distuv"
)

// Holm performs the Holm-Bonferroni step-down procedure.
func Holm[S ~[]E, E any](pValues S, get func(E) float64, alpha float64) S {
	m := len(pValues)
	for i, p := range pValues {
		if get(p)*float64(m-i) > alpha {
			return pValues[:i]
		}
	}
	return pValues
}

// Welch performs the Welch t-test.
// It is ported from the python function pingouin.welch_anova.
func Welch(samples [][]float64) (float64, float64) {
	r := len(samples)
	ddof1 := r - 1
	// log.Printf("ddof1 %d", ddof1)

	var totalN, totalMean float64
	means := make([]float64, 0, r)
	weights := make([]float64, 0, r)
	var adjGrandMean, weightsSum float64
	for _, sample := range samples {
		var mean float64
		for j, s := range sample {
			totalN++
			totalMean += (s - totalMean) / totalN
			mean += (s - mean) / float64(j+1)
		}
		means = append(means, mean)

		w := float64(len(sample)) / stats.Variance(sample)
		weights = append(weights, w)

		adjGrandMean += w * mean
		weightsSum += w
	}
	adjGrandMean /= weightsSum
	// log.Printf("weights: %v, adjGrandMean: %f", weights, adjGrandMean)

	var ssRes, ssBet, ssBetAdj float64
	for i, sample := range samples {
		for _, s := range sample {
			d := s - means[i]
			ssRes += d * d
		}

		d := means[i] - totalMean
		ssBet += d * d * float64(len(sample))

		d = means[i] - adjGrandMean
		ssBetAdj += d * d * weights[i]
	}
	msBetAdj := ssBetAdj / float64(ddof1)
	// log.Printf("ssRes: %f, ssBet: %f, ssBetAdj: %f, msBetAdj: %f", ssRes, ssBet, ssBetAdj, msBetAdj)

	var lamb float64
	for i, sample := range samples {
		cnt := 1 / float64(len(sample)-1)
		mw := 1 - weights[i]/weightsSum
		lamb += cnt * mw * mw
	}
	// log.Printf("lamb0: %f", lamb)
	lamb = 3 * lamb / float64(r*r-1)
	fval := msBetAdj / (1 + (2*lamb*(float64(r)-2))/3)
	// log.Printf("lamb: %f, fval: %f", lamb, fval)
	pval := sf(fval, float64(ddof1), 1/lamb)

	return fval, pval
}

// BrownForsythe performs the Brown-Forsythe test of equal variance.
// The input samples are assumed to be already sorted in ascending order.
// It is ported from the python function scipy.stats.levene.
func BrownForsythe(samples [][]float64) (float64, float64) {
	k := len(samples)
	ni := make([]float64, 0, k)
	yci := make([]float64, 0, k)
	var ntot float64
	for j := 0; j < k; j++ {
		nij := float64(len(samples[j]))
		ni = append(ni, nij)
		yci = append(yci, QuantileF(samples[j], 0.5))
		ntot += nij
	}
	// log.Printf("k: %d, ni: %v, yci: %v, ntot: %f", k, ni, yci, ntot)

	zij := make([][]float64, 0, k)
	for i, sample := range samples {
		ycii := yci[i]

		zi := make([]float64, 0, len(sample))
		for _, s := range sample {
			zi = append(zi, math.Abs(s-ycii))
		}
		zij = append(zij, zi)
	}

	zbari := make([]float64, 0, k)
	var zbar float64
	for i, ziji := range zij {
		zbarii := stats.Mean(ziji)
		zbari = append(zbari, zbarii)
		zbar += zbarii * ni[i]
	}
	zbar /= ntot

	var sumz float64
	for i, zbarii := range zbari {
		d := (zbarii - zbar)
		sumz += ni[i] * d * d
	}
	numer := (ntot - float64(k)) * sumz
	// log.Printf("zbari: %v, zbar: %f, numer: %f", zbari, zbar, numer)

	var dvar float64
	for i, ziji := range zij {
		zbarii := zbari[i]

		for _, zijij := range ziji {
			d := zijij - zbarii
			dvar += d * d
		}
	}
	denom := (float64(k) - 1) * dvar
	// log.Printf("dvar: %f, denom: %f", dvar, denom)

	w := numer / denom
	pval := sf(w, float64(k)-1, ntot-float64(k))
	return w, pval
}

func sf(x, dfn, dfd float64) float64 {
	f := distuv.F{D1: dfn, D2: dfd}
	return f.Survival(x)
}

// Mean returns the mean of a sample
func Mean[S ~[]E, E any](sample S, get func(E) float64) float64 {
	var m float64
	for i, d := range sample {
		m += (get(d) - m) / float64(i+1)
	}
	return m
}

// SS returns the mean and sum of squares of a sample.
func SS[S ~[]E, E any](sample S, get func(E) float64) (float64, float64) {
	mean, M2 := 0.0, 0.0
	for n, d := range sample {
		x := get(d)
		delta := x - mean
		mean += delta / float64(n+1)
		M2 += delta * (x - mean)
	}
	return mean, M2
}

// Quantile returns the q-quantile of a sample.
// The sample is assumed to be already sorted.
// It uses the R8 method described in https://www.itl.nist.gov/div898/handbook/prc/section2/prc262.htm.
func Quantile[S ~[]E, E any](sample S, get func(E) float64, q float64) float64 {
	N := float64(len(sample))
	n := 1/3.0 + q*(N+1/3.0)

	kf, frac := math.Modf(n)
	k := int(kf)
	if k <= 0 {
		return get(sample[0])
	} else if k >= len(sample) {
		return get(sample[len(sample)-1])
	}

	return get(sample[k-1]) + frac*(get(sample[k])-get(sample[k-1]))
}

// QuantileF returns the q-quantile of a sample of float64s.
// The sample is assumed to be already sorted.
func QuantileF(sample []float64, q float64) float64 {
	return Quantile(sample, func(f float64) float64 { return f }, q)
}
